pkg/controller/snapshotter/snapshotter.go

```diff
@@ -160,94 +165,87 @@ func (vs *volumeSnapshotter) getSnapshotDataFromSnapshotName(snapshotName string
 // if needed. Finish waiting when the snapshot becomes available/ready or error.
-func (vs *volumeSnapshotter) waitForPluginSnapshot(snapshotName string, snapshot *crdv1.VolumeSnapshot) error {
-	var snapshotDataObj crdv1.VolumeSnapshotData
-	snapshotDataName := snapshot.Spec.SnapshotDataName
-	glog.Infof("In waitForPluginSnapshot: %s", snapshotName)
-	err := vs.restClient.Get().
-		Name(snapshotDataName).
-		Resource(crdv1.VolumeSnapshotDataResourcePlural).
-		Namespace(v1.NamespaceDefault).
-		Do().Into(&snapshotDataObj)
-	if err != nil {
+func (vs *volumeSnapshotter) waitForSnapshot(snapshotName string, snapshot *crdv1.VolumeSnapshot) error {
+        var snapshotDataObj crdv1.VolumeSnapshotData
+        snapshotDataName := snapshot.Spec.SnapshotDataName
+        glog.Infof("In waitForSnapshot: %s", snapshotName)
+        err := vs.restClient.Get().
+                Name(snapshotDataName).
+                Resource(crdv1.VolumeSnapshotDataResourcePlural).
+                Namespace(v1.NamespaceDefault).
+                Do().Into(&snapshotDataObj)
+        if err != nil {
+		glog.Infof("waitForSnapshot: %s. VolumeSnapshotData not created yet.", snapshotName)
 		return err
 	}
 
-	pv, err := vs.getPVFromVolumeSnapshot(snapshotName, snapshot)
-	if err != nil {
-		return err
-	}
-	spec := &pv.Spec
-	volumeType := crdv1.GetSupportedVolumeFromPVSpec(spec)
-	if len(volumeType) == 0 {
-		return fmt.Errorf("unsupported volume type found in PV %#v", spec)
-	}
-	plugin, ok := (*vs.volumePlugins)[volumeType]
-	if !ok {
-		return fmt.Errorf("%s is not supported volume for %#v", volumeType, spec)
-	}
+	plugin, err := vs.getPlugin(snapshotName, snapshot)
+        if plugin == nil {
+		return fmt.Errorf("Failed to get volume plugin to wait for snapshot creation %s. %v", snapshotName, err)
+        }
 
-	var completed bool = false
 	var conditions *[]crdv1.VolumeSnapshotCondition = nil
+	var status, newstatus string = "", ""
 	// Wait until the snapshot is successfully created by the plugin or an error occurs that
 	// fails the snapshot creation.
 	for {
-		// TODO(xyang): Use status returned from DescribeSnapshot to update snapshot
-		conditions, completed, err = plugin.DescribeSnapshot(&snapshotDataObj)
-		if !completed {
-			glog.Infof("Snapshot %s creation is not complete yet. Status: [%#v] Retrying...", snapshotName, conditions)
-			if err != nil {
-				// Don't bail out here. The error message could be "current snapshot status is: creating".
-				glog.Infof("Message received when creating snapshot %s: %v", snapshotName, err)
+		conditions, _, err = (*plugin).DescribeSnapshot(&snapshotDataObj)
+		newstatus = vs.getSimplifiedSnapshotStatus(conditions)
+		if newstatus == statusPending {
+			if newstatus != status {
+				status = newstatus
+				glog.Infof("Snapshot %s creation is not complete yet. Status: [%#v] Retrying...", snapshotName, conditions)
+				if err != nil {
```

Comment:
I think the err check should be outside?

Reply:
Sure.  I think the logic of DescribeSnapshot needs to be changed.  Right now if the snapshot is in creating status in cinder, there is an error "current snapshot status is: creating".